command/ca/certificate.go

```diff
@@ -443,3 +486,408 @@ func parseTimeDuration(ctx *cli.Context) (notBefore api.TimeDuration, notAfter a
+	srv := &http.Server{Addr: addr}
+
+	http.HandleFunc(fmt.Sprintf("/.well-known/acme-challenge/%s", token), func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(keyAuth))
+	})
+
+	go func() {
+		// returns ErrServerClosed on graceful close
+		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+			// NOTE: there is a chance that next line won't have time to run,
+			// as main() doesn't wait for this goroutine to stop. don't use
+			// code with race conditions like these for production. see post
+			// comments below on more discussion on how to handle this.
+			ui.Printf("\nListenAndServe(): %s\n", err)
+		}
+	}()
+
+	// returning reference so caller can call Shutdown()
+	return srv
+}
+
+type issueMode interface {
+	Run() error
+	Cleanup() error
+}
+
+type standaloneMode struct {
+	identifier, token string
+	key               *jose.JSONWebKey
+	port              int
+	srv               *http.Server
+}
+
+func newStandaloneMode(identifier string, port int, token string, key *jose.JSONWebKey) *standaloneMode {
+	return &standaloneMode{
+		identifier: identifier,
+		port:       port,
+		token:      token,
+		key:        key,
+	}
+}
+
+func (sm *standaloneMode) Run() error {
+	ui.Printf("Using Standalone Mode HTTP challenge to validate %s .", sm.identifier)
+	keyAuth, err := acme.KeyAuthorization(sm.token, sm.key)
+	if err != nil {
+		return errors.Wrap(err, "error generating ACME key authorization")
+	}
+	sm.srv = startHTTPServer(fmt.Sprintf("0.0.0.0:%d", sm.port), sm.token, keyAuth)
+	return nil
+}
+
+func (sm *standaloneMode) Cleanup() error {
+	return errors.Wrap(sm.srv.Shutdown(context.TODO()), "error gracefully shutting down server")
+}
+
+type webrootMode struct {
+	dir, token, identifier string
+	key                    *jose.JSONWebKey
+}
+
+func newWebrootMode(dir, token, identifier string, key *jose.JSONWebKey) *webrootMode {
+	return &webrootMode{
+		dir:        dir,
+		token:      token,
+		identifier: identifier,
+		key:        key,
+	}
+}
+
+func (wm *webrootMode) Run() error {
+	ui.Printf("Using Webroot Mode HTTP challenge to validate %s .", wm.identifier)
+	keyAuth, err := acme.KeyAuthorization(wm.token, wm.key)
+	if err != nil {
+		return errors.Wrap(err, "error generating ACME key authorization")
+	}
+	_, err = os.Stat(wm.dir)
+	switch {
+	case os.IsNotExist(err):
+		return errors.Errorf("webroot directory %s does not exist", wm.dir)
+	case err != nil:
+		return errors.Wrapf(err, "error checking for directory %s", wm.dir)
+	}
+
+	chPath := fmt.Sprintf("%s/.well-known/acme-challenge", wm.dir)
+	if _, err = os.Stat(chPath); os.IsNotExist(err) {
+		if err = os.MkdirAll(chPath, 0700); err != nil {
+			return errors.Wrapf(err, "error creating directory path %s", chPath)
+		}
+	}
+
+	return errors.Wrapf(ioutil.WriteFile(fmt.Sprintf("%s/%s", chPath, wm.token), []byte(keyAuth), 0600),
+		"error writing key authorization file %s", chPath+wm.token)
+}
```

Comment:
Are you saying I should add more documentation to explain this better?

Reply:
The webroot usage was just a copy&paste of `--standalone`, so yes, we need proper usage docs, but at the same time we need to document, or add proper permissions to files so they can be read.
For example, on many systems nginx runs on the user `www-data`, so if you create a file/directory with 0700 it will be only accessible if you're running step with the same user, or if you `chown` later. We can check what other utils do, do they use more permissive perms (755 for directories, 644 for files), or is the user who has to chown the files?

Reply:
Just checked acme.sh and lego.

* lego attempts to write both the directory and the challenge with permissive perms (755/644) as you said.
* acme attempts to write the file and then chown it to the user that owned the root website directory.

Either way you'd proably need `sudo`. I think using permissive perms is a bit cleaner. Thoughts?

Reply:
Yes I think 755/644 is probably the best solution here